File: netlogo-gui/src/main/prim/_repeatinternal.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.api.Let;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Command;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.MutableLong;

public final strictfp class _repeatinternal
    extends Command {
  private final Let let;

  public _repeatinternal(int offset, Let let) {
    this.offset = offset;
    this.let = let;
  }

  @Override
  public Syntax syntax() {
    return Syntax.commandSyntax();
  }

  @Override
  public String toString() {
    return super.toString() + ":" + offset;
  }

  @Override
  public void perform(final Context context) {
    MutableLong counter = (MutableLong) context.getLet(let);
    if (counter.value() <= 0) {
      context.ip = next;
    } else {
      counter.value_$eq(counter.value() - 1);
      context.ip = offset;
    }
  }

  public void perform_1(final Context context) {
    MutableLong counter = (MutableLong) context.getLet(let);
    if (counter.value() <= 0) {
      context.ip = next;
    } else {
      counter.value_$eq(counter.value() - 1);
      context.ip = offset;
    }
  }
}

File: netlogo-gui/src/main/prim/_repeatinternal.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.core.{ Let, Syntax }
import org.nlogo.nvm.{ Command, Context, MutableLong }

class _repeatinternal(_offset: Int, private[this] val _let: Let) extends Command {
  offset = _offset
  def let = _let

  override def toString: String =
    super.toString + ":" + offset

  override def perform(context: Context): Unit = {
    val counter = context.activation.binding.getLet(_let).asInstanceOf[MutableLong]
    if (counter.value <= 0) {
      context.ip = next
    } else {
      counter.value = counter.value - 1
      context.ip = offset
    }
  }

  def perform_1(context: Context): Unit = {
    val counter = context.activation.binding.getLet(_let).asInstanceOf[MutableLong]
    if (counter.value <= 0) {
      context.ip = next
    } else {
      counter.value = counter.value - 1
      context.ip = offset
    }
  }
}
